### PR TITLE
Markup adjustments & expose `withCell` for custom Cell components

### PR DESCRIPTION
**Pull request by sophiabits**
Our use case requires sticky headers, as we're embedding the sheet inside text posts. The sheets are also reasonably large and so for performance reasons we want to virtualize the cells. So we need to use divs instead of table markup.

For the most part this works pretty well out of the box! But there are some things which need to be fixed:

- [X] Column indicator row is wrapped with a <tr> instead of using the `Row` component.
- [ ] The "corner" cell that gets rendered when both column and row indicators are enabled is rendered as a <th>. I think the best way to fix this is to add a new `CornerIndicator` prop which would be rendered where the th currently is. Rendering a <ColumnIndicator> or <RowIndicator> would be a breaking change here, and that corner piece is often used for special functionality anyway (for instance in Excel, clicking the corner indicator selects all of your data)
- [ ] We need some way of accessing the cell actions within our custom `Cell` component. There's a todo item in Spreadsheet.js which indicates that the plan is to automatically enhance the component passed in by the user. If that's the way you want to do it, I can refactor my branch to behave that way but I think it makes more sense to export the `enhanceCell` HOC directly so I've done it that way. Let me know what you prefer 👍 
    - [ ] I'm not very good with flow, but do we need to export any new types because of this? I don't think any types get exported atm aside from the main SpreadsheetStateProvider's prop types but I could be missing something.
	- [ ] If #72 gets merged, then `enhanceCell` will require its types to be defined.
- [ ] Is it worth adjusting the base css to be less table-centric (e.g. `.Spreadsheet td` --> `.Spreadsheet .SpreadsheetCell` or similar)? I think it's probably fine as-is.

If you can let me know how you'd like those above questions resolved, I'll finish off the PR ^^ I'm just not comfortable making API changes to the project on my own. But if we can get this done then it'll be a lot easier for people to add more complex behavior to their sheets :)

### CODE REVIEW

**Comment by iddan** [comment]
I agree a `CornerIndicator` is a good idea

**Comment by iddan** [comment]
> Is it worth adjusting the base css to be less table-centric
Yes, but better use BEM `.Spreadsheet__cell`

**Comment by iddan** [comment]
| We need some way of accessing the cell actions within our custom Cell component. There's a todo item in Spreadsheet.js which indicates that the plan is to automatically enhance the component passed in by the user. If that's the way you want to do it, I can refactor my branch to behave that way but I think it makes more sense to export the enhanceCell HOC directly so I've done it that way. Let me know what you prefer

I'd prefer to enhance the incoming Cell component. It's a better API as in the future all the HOCs should be refactored out. If you can update the PR that would be lovely thank you

File: src/CornerIndicator.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
+// @flow
+
+export type Props = {};
```
Comment (iddan) [inline src/CornerIndicator.js:3]: I think this should the props of th element

File: __fixtures__/WIthCornerIndicator.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
+import { createFixture } from "react-cosmos";
```
Comment (iddan) [inline __fixtures__/WIthCornerIndicator.js:1]: The filename should be WithCornerIndicator.js

File: src/Spreadsheet.js [hunk 3]
```diff
@@ -252,11 +252,12 @@ class Spreadsheet<CellType, Value> extends PureComponent<{|
     this.root = root;
   };
 
+  getCellComponent = memoizeOne(enhanceCell);
```
Comment (iddan) [inline src/Spreadsheet.js:255]: I don't under this.

Comment (sophiabits) [reply src/Spreadsheet.js:255]: It's memoizing the application of `enhanceCell` to the `Cell`  prop. If `Cell` changes between renders, the new component will be enhanced and that enhanced component will be returned from `getCellComponent` every time it gets called (unless `Cell` changes again)

In a functional component we'd be able to use hooks like so:

```js
function Spreadsheet({ Cell, ... }) {
  const ConnectedCell = useMemo(() => enhanceCell(Cell), [Cell]);
  // now we can use ConnectedCell
}
```

If that makes it a bit clearer.

Comment (iddan) [reply src/Spreadsheet.js:255]: Great. Can you add a short doc comment to explain this

Comment (sophiabits) [reply src/Spreadsheet.js:255]: Will do :)

**Comment by iddan** [comment]
I think the branch is behind `master`. Can you merge the latest to this branch so the diff will make sense?

**Comment by sophiabits** [comment]
I think that's it

**Comment by iddan** [comment]
Thank you very much for your contribution

**Comment by iddan** [comment]
I have an issue with publishing to npm. I contacted support

**Comment by iddan** [comment]
Published in v0.4.47

**Pull request merged**